test/example.c

```diff
@@ -19,13 +19,108 @@
 #  define TESTFILE "foo.gz"
 #endif
 
-#define CHECK_ERR(err, msg) { \
-    if (err != Z_OK) { \
-        fprintf(stderr, "%s error: %d\n", msg, err); \
-        exit(1); \
+#define SUCCESSFUL                 1
+#define FAILED_WITH_ERROR_CODE     0
+#define FAILED_WITHOUT_ERROR_CODE -1
+
+typedef struct test_result_s {
+    int         result; /* One of: SUCCESSFUL,
+                           FAILED_WITH_ERROR_CODE, or
+                           FAILED_WITHOUT_ERROR_CODE*/
+    int         error_code; /* error code if success is FAILED_WITH_ERROR_CODE */
+    int         line_number;
+    const char* message;
+    const char* extended_message;
+} test_result;
+
+#define STRING_BUFFER_SIZE 100
+char string_buffer[STRING_BUFFER_SIZE];
+
+#define CHECK_ERR(_error_code, _message) { \
+    if (_error_code != Z_OK) { \
+        test_result result; \
+        result.result = FAILED_WITH_ERROR_CODE; \
+        result.error_code = _error_code; \
+        result.line_number = __LINE__; \
+        result.message = _message; \
+        result.extended_message = NULL; \
+        return result; \
     } \
 }
 
+#define RETURN_FAILURE(_message, _extended_message) { \
+    test_result result; \
+    result.result = FAILED_WITHOUT_ERROR_CODE; \
+    result.error_code = Z_OK; \
+    result.line_number = __LINE__; \
+    result.message = _message; \
+    result.extended_message = _extended_message; \
+    return result; \
+}
+
+#define RETURN_SUCCESS(_message, _extended_message) { \
+    test_result result; \
+    result.result = SUCCESSFUL; \
+    result.error_code = Z_OK; \
+    result.message = _message; \
+    result.extended_message = _extended_message; \
+    return result; \
+}
+
+void handle_test_results OF((FILE* output, test_result result, z_const char* testcase_name, int is_junit_output, int* failed_test_count));
+
+void handle_test_results(output, result, testcase_name, is_junit_output, failed_test_count)
```

Comment:
I found handle_test_results fragile in that it was hard to change the XML format because the two output formats and the state management were all mixed together.  (Hence the commit in my pull request to refactor it.)

Reply:
That's what I meant by "formatter" earlier.
We could produce multiple formats of output rather than run the tests multiple times.

That said, I think you're right, that I should have two separate functions for the two formats.
I'll do that soon.